go-selinux/selinux_linux.go

```diff
@@ -445,6 +467,217 @@ func ComputeCreateContext(source string, target string, class string) (string, e
 	return readWriteCon(filepath.Join(getSelinuxMountPoint(), "create"), fmt.Sprintf("%s %s %d", source, target, classidx))
 }
 
+// catsToBitset stores categories in a bitset.
+func catsToBitset(cats string) (*bitset.BitSet, error) {
+	bitset := &bitset.BitSet{}
+
+	catlist := strings.Split(cats, ",")
+	for _, r := range catlist {
+		ranges := strings.SplitN(r, ".", 2)
+		if len(ranges) > 1 {
+			catstart, err := parseLevelItem(ranges[0], category)
+			if err != nil {
+				return nil, err
+			}
+			catend, err := parseLevelItem(ranges[1], category)
+			if err != nil {
+				return nil, err
+			}
+			for i := catstart; i <= catend; i++ {
+				bitset.Set(i)
+			}
+		} else {
+			cat, err := parseLevelItem(ranges[0], category)
+			if err != nil {
+				return nil, err
+			}
+			bitset.Set(cat)
+		}
+	}
+
+	return bitset, nil
+}
+
+// parseLevelItem parses and verifies that a sensitivity or category are valid
+func parseLevelItem(s string, sep levelItem) (uint, error) {
+	if levelItem(s[0]) != sep {
+		return 0, ErrLevelSyntax
+	}
+	val, err := strconv.ParseUint(s[1:], 10, 32)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to parse level")
+	}
+
+	return uint(val), nil
+}
+
+// rangeStrToMLSRange marshals a string representation of a range.
+func rangeStrToMLSRange(rangeStr string) (*mlsRange, error) {
+	mlsRange := &mlsRange{}
+	levelSlice := strings.SplitN(rangeStr, "-", 2)
+
+	switch len(levelSlice) {
+	case 2:
+		high := strings.SplitN(levelSlice[1], ":", 2)
+		sensHigh, err := parseLevelItem(high[0], sensitivity)
+		if err != nil {
+			return nil, err
+		}
+		mlsRange.high = &level{sens: sensHigh}
+		if len(high) > 1 {
+			catsHigh, err := catsToBitset(high[1])
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to parse categories")
+			}
+			mlsRange.high.cats = catsHigh
+		}
+		fallthrough
+	case 1:
+		low := strings.SplitN(levelSlice[0], ":", 2)
+		sensLow, err := parseLevelItem(low[0], sensitivity)
+		if err != nil {
+			return nil, err
+		}
+		mlsRange.low = &level{sens: sensLow}
+		if len(low) > 1 {
+			catsLow, err := catsToBitset(low[1])
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to parse categories")
+			}
+			mlsRange.low.cats = catsLow
+		}
+	}
+
+	if mlsRange.high == nil {
+		mlsRange.high = mlsRange.low
+	}
+
+	return mlsRange, nil
+}
+
+/*
+bitsetToStr takes a category bitset and returns it in the
+canonical selinux syntax
+*/
```

Comment:
nit: maybe use `//`-style comment here. I'm not exactly sure why but `/* ... */` is not widely popular/used in Go

Reply:
done